server/tga/sign_off/utils.py

```diff
@@ -0,0 +1,124 @@
+from typing import Dict, List, Optional, Any
+import pathlib
+from copy import deepcopy
+import re
+from time import time
+
+from authlib.jose import JsonWebToken
+from bson import ObjectId
+from bson.errors import InvalidId
+from flask import g, current_app as app
+from flask_babel import _
+
+from superdesk import get_resource_service
+from superdesk.errors import SuperdeskApiError
+from superdesk.utc import utcnow
+from superdesk.notification import push_notification
+
+from .form import UserSignOffForm
+
+
+JWT_ALGORITHM = "HS256"
+
+
+def get_css_filename():
+    try:
+        dist_path = pathlib.Path(pathlib.Path(__file__).parent.parent.parent.parent, "client/dist")
+        css_filename = [f.name for f in dist_path.glob("app.bundle.*.css")][0]
+        return f"/{css_filename}"
+    except (IndexError, ValueError):
+        return "/app.bundle.css"
+
+
+def modify_asset_urls(item, author_id: ObjectId):
+    pattern = re.compile('\/api\/upload-raw\/(.*?)"')
+    new_body_html = item["body_html"]
+    url_prefix = "/api/upload-raw/"
+    url_prefix_len = len(url_prefix)
+
+    for asset_filename in re.findall(pattern, item["body_html"]):
+        asset_token = gen_jwt_for_approval_request(asset_filename, author_id, "upload-raw", token_expiration=3600)
+        new_body_html = new_body_html.replace(
+            url_prefix + asset_filename, f"/api/sign_off_requests/upload-raw/{asset_token}"
+        )
+
+    item["body_html"] = new_body_html
+
+    for key, association in (item.get("associations") or {}).items():
+        for size_name, rendition in (association.get("renditions") or {}).items():
+            rendition_href = rendition["href"]
+            asset_filename = rendition_href[rendition_href.index(url_prefix) + url_prefix_len :]
+            asset_token = gen_jwt_for_approval_request(asset_filename, author_id, "upload-raw", token_expiration=3600)
+            rendition["href"] = rendition_href.replace(
+                url_prefix + asset_filename, f"/api/sign_off_requests/upload-raw/{asset_token}"
+            )
+
+
+def update_item_publish_approval(item: Dict[str, Any], form: UserSignOffForm):
+    data = deepcopy(item.get("extra") or {})
+    data["publish_sign_off"] = {
```

Comment:
I wonder if it's enough to support only single author?

Reply:
The current implementation is single sign off, and the tickets never mentioned multi sign off.
I was going to discuss with Aleksa on this, and have a new ticket if needed

Reply:
I see, ok then